File: services/api-axum/src/compliance.rs
//! Industry compliance profile, read at startup.
//! One repo serves every industry; COMPLIANCE_PROFILE flips a few controls.
//!
//! Source of truth is `compliance/profiles.json` (relative to the working dir),
//! a generated catalog of 30 profiles sharing the same control keys.

use serde::Serialize;
use serde_json::{Map, Value};
use std::sync::OnceLock;

/// The active profile, shaped for the GET /compliance response.
#[derive(Serialize, Clone)]
pub struct Controls {
    pub profile: String,
    pub name: String,
    pub jurisdiction: String,
    pub controls: Map<String, Value>,
}

impl Controls {
    /// Session length in seconds, read from the active profile's controls.
    /// Falls back to 8 hours when the control is absent or non-numeric.
    pub fn session_timeout_seconds(&self) -> u64 {
        self.controls
            .get("session_timeout_seconds")
            .and_then(Value::as_u64)
            .unwrap_or(8 * 60 * 60)
    }
}

static ACTIVE: OnceLock<Controls> = OnceLock::new();

/// The profile this process booted with.
pub fn active() -> &'static Controls {
    ACTIVE.get_or_init(load)
}

fn load() -> Controls {
    let profile = std::env::var("COMPLIANCE_PROFILE").unwrap_or_else(|_| "baseline".into());
    match load_profile(&profile) {
        Ok(c) => c,
        Err(e) => {
            // A named profile we cannot load must fail loud rather than serve
            // a silently-wrong control set.
            eprintln!("FATAL: compliance profile not loadable: {profile}: {e}");
            std::process::exit(1);
        }
    }
}

/// Read `compliance/profiles.json` and return the selected profile.
/// Kept separate from `load` so tests can drive it without exiting the process.
pub fn load_profile(profile: &str) -> Result<Controls, String> {
    let path = std::path::Path::new("compliance").join("profiles.json");
    let text = std::fs::read_to_string(&path)
        .map_err(|e| format!("cannot read {}: {e}", path.display()))?;
    let doc: Value = serde_json::from_str(&text).map_err(|e| format!("invalid JSON: {e}"))?;

    let profiles = doc
        .get("profiles")
        .and_then(Value::as_object)
        .ok_or_else(|| "missing \"profiles\" object".to_string())?;

    let p = profiles
        .get(profile)
        .and_then(Value::as_object)
        .ok_or_else(|| format!("unknown profile: {profile}"))?;

    let name = p
        .get("name")
        .and_then(Value::as_str)
        .unwrap_or(profile)
        .to_string();
    let jurisdiction = p
        .get("jurisdiction")
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_string();
    let controls = p
        .get("controls")
        .and_then(Value::as_object)
        .ok_or_else(|| format!("profile {profile} has no \"controls\" object"))?
        .clone();

    Ok(Controls {
        profile: profile.to_string(),
        name,
        jurisdiction,
        controls,
    })
}

File: services/protocol-grpc/internal/compliance/compliance.go
// Package compliance reads the active industry profile at startup.
// One repo serves every industry; COMPLIANCE_PROFILE flips a set of controls.
//
// Profiles live in compliance/profiles.json (generated, catalog-versioned).
// Every profile shares the same control keys; only the values differ.
package compliance

import (
	"encoding/json"
	"log"
	"os"
	"path/filepath"
)

// Controls is the active profile resolved at boot.
//
// Controls holds the raw per-control map plus a few derived fields the rest
// of the service reads directly (for example session length).
type Controls struct {
	Profile      string                 `json:"profile"`
	Name         string                 `json:"name"`
	Jurisdiction string                 `json:"jurisdiction"`
	Controls     map[string]interface{} `json:"controls"`

	// SessionTimeoutSeconds is derived from controls["session_timeout_seconds"].
	// Default 8h when the control is absent.
	SessionTimeoutSeconds int `json:"-"`
}

// profilesFile is the on-disk shape of compliance/profiles.json.
type profilesFile struct {
	Device         string                 `json:"device"`
	CatalogVersion int                    `json:"catalogVersion"`
	Profiles       map[string]profileSpec `json:"profiles"`
}

type profileSpec struct {
	Name         string                 `json:"name"`
	Priority     string                 `json:"priority"`
	Jurisdiction string                 `json:"jurisdiction"`
	Controls     map[string]interface{} `json:"controls"`
}

// Active is the profile this process booted with.
var Active = load()

func load() Controls {
	profile := envOr("COMPLIANCE_PROFILE", "baseline")

	// Safe default returned when the catalog cannot be read at all. This keeps
	// importing packages (whose working dir is not the service root, e.g. test
	// binaries) loadable; the service itself runs from the root where the file
	// resolves and a missing/garbled catalog there is a real fail-loud below.
	fallback := Controls{Profile: profile, SessionTimeoutSeconds: 8 * 60 * 60, Controls: map[string]interface{}{}}

	wd, _ := os.Getwd()
	raw, err := os.ReadFile(filepath.Join(wd, "compliance", "profiles.json"))
	if err != nil {
		if os.IsNotExist(err) {
			return fallback
		}
		log.Fatalf("FATAL: compliance profiles not loadable: %v", err)
	}

	var pf profilesFile
	if err := json.Unmarshal(raw, &pf); err != nil {
		log.Fatalf("FATAL: compliance profiles unparseable: %v", err)
	}

	spec, ok := pf.Profiles[profile]
	if !ok {
		log.Fatalf("FATAL: unknown COMPLIANCE_PROFILE: %s", profile)
	}

	c := Controls{
		Profile:               profile,
		Name:                  spec.Name,
		Jurisdiction:          spec.Jurisdiction,
		Controls:              spec.Controls,
		SessionTimeoutSeconds: 8 * 60 * 60,
	}

	// Derive session length from the control map when present.
	if v, ok := spec.Controls["session_timeout_seconds"]; ok {
		if n, ok := v.(float64); ok && n > 0 {
			c.SessionTimeoutSeconds = int(n)
		}
	}

	return c
}

func envOr(k, d string) string {
	if v := os.Getenv(k); v != "" {
		return v
	}
	return d
}
